### PR TITLE
[CPU] Added separate inference for pc, node splitting

### PR DESCRIPTION
**Pull request by a-sidorova**
### Details:
 - *Added separate variables for constant and inconstant nodes to avoid conditions in inference cycle*
 - *Added separate method for inference with pc*

### Tickets:
 - *56304, 56308*

### CODE REVIEW

**Comment by a-sidorova** [comment]
@maxnick could you take a look please?

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 1]
```diff
@@ -390,6 +392,16 @@ void MKLDNNGraph::InitOptimalPrimitiveDescriptors() {
     }
 }
 
+void MKLDNNGraph::SplitNodes() {
+    OV_ITT_SCOPE(FIRST_INFERENCE, itt::domains::MKLDNN_LT, "MKLDNNGraph::SplitNodes");
+    for (auto& graphNode : graphNodes) {
+        if (graphNode->isConstant())
+            constantGraphNodes.emplace_back(graphNode);
+        else
+            inconstantGraphNodes.emplace_back(graphNode);
```
Comment (maxnick) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:401]: M.b. rename to `mutableGraphNodes` or `regularGraphNodes`.  What do you think?

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 2]
```diff
@@ -806,28 +815,73 @@ void MKLDNNGraph::Infer(MKLDNNInferRequest* request, int batch) {
...
+        OV_ITT_SCOPED_TASK(itt::domains::MKLDNNPlugin, inconstantGraphNodes[i]->profiling.execute);
+        inconstantGraphNodes[i]->execute(stream);
+
+        ENABLE_CPU_DEBUG_CAP(nd.dumpOutputBlobs(inconstantGraphNodes[i]));
+    }
+
+    if (infer_count != -1) infer_count++;
+}
+
+void MKLDNNGraph::InferWithPerfCounters(MKLDNNInferRequest* request, int batch) {
```
Comment (maxnick) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:855]: Looks like there is a lot of code duplication with `Infer()` method. Not good.

**Review by maxnick — CHANGES_REQUESTED** [review]

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 1 above)
Comment (a-sidorova) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp]: I think `constantGraphNodes` is clear but I agree that it is worth renaming `inconstantGraphNodes`

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 2 above)
Comment (a-sidorova) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp]: I wanted to avoid frequent conditions in cycle
But I have checked It seems like there is no overhead. So I added If statement
Thank you!

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 3]
```diff
@@ -810,24 +819,33 @@ void MKLDNNGraph::Infer(MKLDNNInferRequest* request, int batch) {
...
-    for (int i = 0; i < graphNodes.size(); i++) {
-        if (request != nullptr) {
+    for (int i = 0; i < constantGraphNodes.size(); i++) {
+        if (request != nullptr)
+            request->ThrowIfCanceled();
+
+        if (batch > 0)
+            constantGraphNodes[i]->setDynamicBatchLim(batch);
+
+        ENABLE_CPU_DEBUG_CAP(nd.dumpOutputBlobs(constantGraphNodes[i]));
```
Comment (maxnick) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:829]: I think we should also dump input  blobs for the constant nodes as there might be intermediate layers on a constant path such as reorder, reshape etc.

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 4]
```diff
@@ -810,24 +819,33 @@ void MKLDNNGraph::Infer(MKLDNNInferRequest* request, int batch) {
...
     ENABLE_CPU_DEBUG_CAP(NodeDumper nd(config.debugCaps, infer_count));
 
-    for (int i = 0; i < graphNodes.size(); i++) {
-        if (request != nullptr) {
+    for (int i = 0; i < constantGraphNodes.size(); i++) {
+        if (request != nullptr)
+            request->ThrowIfCanceled();
+
+        if (batch > 0)
+            constantGraphNodes[i]->setDynamicBatchLim(batch);
```
Comment (maxnick) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:827]: Does this call make sense for a constant node?

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 3 above)
Comment (a-sidorova) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:828]: Yes, I added, thanks

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 4 above)
Comment (a-sidorova) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp]: Yes, it's useless for constant nodes, I removed.
And now for constant layer we only dump blobs so I added directive `ifdef` with `CPU_DEBUG_CAPS` for this

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 3 above)
Comment (maxnick) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:828]: Great! But why do we need this check `if (request != nullptr)` ?

**Review by maxnick — APPROVED** [review]

File: inference-engine/src/mkldnn_plugin/perf_count.h [hunk 5]
```diff
@@ -46,4 +46,4 @@ class PerfHelper {
 
 }  // namespace MKLDNNPlugin
 
-#define PERF(_counter) PerfHelper __helper##__counter (_counter->PerfCounter());
+#define PERF(_counter) std::unique_ptr<PerfHelper>(new PerfHelper(_counter->PerfCounter()))
```
Comment (maxnick) [inline inference-engine/src/mkldnn_plugin/perf_count.h:49]: BTW. Do we really need this macro?

Comment (a-sidorova) [reply inference-engine/src/mkldnn_plugin/perf_count.h:49]: I think, it looks better when all code about PerfCount is in `perf_count.h` and user shouldn't know how macro `PERF` works. It just calculates perfcounters. May be it's a matter of taste :blush:

Thank you!

**Comment by a-sidorova** [comment]
@dmitry-gorokhov could you take a look please?

File: inference-engine/src/mkldnn_plugin/perf_count.h (hunk 5 above)
Comment (maxnick) [reply inference-engine/src/mkldnn_plugin/perf_count.h:49]: But now the user must prepare the std::unique_ptr<PerfHelper> in order to use this functionality. So it doesn't look like an full encapsulation in this macro.

**Comment by maxnick** [comment]
I just remembered yet another important thing that I missed earlier. In the previous implementation, we allocated `PerfHelper` object on stack with minimal additional overhead on allocation (just rsp decrease), but now, due to the use of the unique ptr, we allocate the `PerfHelper` object on the heap, which means additional malloc call on every iteration, which significantly increases the overhead cost versus stack allocation. The question is whether we can afford such additional overhead for time measurement  in a real application, because although malloc is more expensive than  stack allocation, it can still be negligible compared to the entire node execution procedure.

**Comment by a-sidorova** [comment]
> I just remembered yet another important thing that I missed earlier. In the previous implementation, we allocated `PerfHelper` object on stack with minimal additional overhead on allocation (just rsp decrease), but now, due to the use of the unique ptr, we allocate the `PerfHelper` object on the heap, which means additional malloc call on every iteration, which significantly increases the overhead cost versus stack allocation. The question is whether we can afford such additional overhead for time measurement in a real application, because although malloc is more expensive than stack allocation, it can still be negligible compared to the entire node execution procedure.

@maxnick good catch, thank you! fps drop between case with pc and case without pc is near 5%
@dmitry-gorokhov is this overhead ok?

**Comment by maxnick** [comment]
This 5% drop is due to many things, not just memory allocation, which is the less significant part.

**Comment by dmitry-gorokhov** [comment]
> This 5% drop is due to many things, not just memory allocation, which is the less significant part.

**Comment by dmitry-gorokhov** [comment]
> This 5% drop is due to many things, not just memory allocation, which is the less significant part.

@a-sidorova  @maxnick  what overhead do we head on master right now? I expect it shouldn't become bigger because of these changes.

**Comment by a-sidorova** [comment]
@dmitry-gorokhov on master we work with object of class [PerfHelper](https://github.com/openvinotoolkit/openvino/blob/8fb1182670a0d6d49858239b6b4f387d8de6074d/inference-engine/src/mkldnn_plugin/perf_count.h#L38). But in this PR we work with unique_ptr for this object. So by the overhead we mean allocation on heap for ptr. But I think that it's minor overhead

**Comment by dmitry-gorokhov** [comment]
> @dmitry-gorokhov on master we work with object of class [PerfHelper](https://github.com/openvinotoolkit/openvino/blob/8fb1182670a0d6d49858239b6b4f387d8de6074d/inference-engine/src/mkldnn_plugin/perf_count.h#L38). But in this PR we work with unique_ptr for this object. So by the overhead we mean allocation on heap for ptr. But I think that it's minor overhead

I would rephrase. By introducing unique_ptr approach don't we degrade perf counters values for specific usecases (like very small models) comparing with master?

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 6]
```diff
@@ -810,24 +819,34 @@ void MKLDNNGraph::Infer(MKLDNNInferRequest* request, int batch) {
...
+#endif
+
+    std::unique_ptr<PerfHelper> perf;
+    const bool collectPerfCounters = config.collectPerfCounters;
+    for (int i = 0; i < mutableGraphNodes.size(); i++) {
+        if (request != nullptr)
+            request->ThrowIfCanceled();
+
+        if (collectPerfCounters)
+            perf = PERF(mutableGraphNodes[i]);
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:839]: If I got it right with this change we start including `ThrowIfCanceled()` into pc (but from the next iteration). In that case I want to understand how perf counters are affected? It also makes sense to move `perf` initialization in the very beginning of the loop.

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.h [hunk 7]
```diff
@@ -218,6 +221,7 @@ class MKLDNNGraph {
     void AllocateWithReuse();
     void CreatePrimitives();
     void ExecuteConstantNodesOnly();
+    void SplitNodes();
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.h:224]: The method has very generic name. Maybe rename to "ExtractConstantNodes"? @maxnick what do you think?

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.h [hunk 8]
```diff
@@ -200,6 +200,9 @@ class MKLDNNGraph {
     std::vector<MKLDNNNodePtr> graphNodes;
     std::vector<MKLDNNEdgePtr> graphEdges;
 
+    std::vector<MKLDNNNodePtr> constantGraphNodes;
+    std::vector<MKLDNNNodePtr> mutableGraphNodes;
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.h:204]: The problem I see with this approach: we create duplicate holders for graph nodes (in addition two graphNodes) which are not synchronized in any way. Do we have any good approach two encapsulate them somewhere to prevent others from using them (before SplitNodes() call)?
From the top of my mind I can suggest remove them from fields and pass to required functions as parameters. 
Otherwise we at least have to move them into private section and add detailed comment with usage insights. 

@maxnick any thoughts?

**Comment by EgorDuplensky** [comment]
> > I just remembered yet another important thing that I missed earlier. In the previous implementation, we allocated `PerfHelper` object on stack with minimal additional overhead on allocation (just rsp decrease), but now, due to the use of the unique ptr, we allocate the `PerfHelper` object on the heap, which means additional malloc call on every iteration, which significantly increases the overhead cost versus stack allocation. The question is whether we can afford such additional overhead for time measurement in a real application, because although malloc is more expensive than stack allocation, it can still be negligible compared to the entire node execution procedure.
> 
> @maxnick good catch, thank you! fps drop between case with pc and case without pc is near 5%
> @dmitry-gorokhov is this overhead ok?

@maxnick @a-sidorova  One of the PRs I am working on is also updating Infer() method.
Can we do the following instead of unique_ptr?:
```c++
inline void MKLDNNGraph::executeNode(const MKLDNNNodePtr& node, const mkldnn::stream& stream) const {
    auto execute = [](const MKLDNNNodePtr& node, const mkldnn::stream& s) {
        OV_ITT_SCOPED_TASK(itt::domains::MKLDNNPlugin, node->profiling.execute);
        node->execute(s);
    };

    auto executeAndMeasure = [](const MKLDNNNodePtr& node, const mkldnn::stream& s) {
        PERF(node);
        OV_ITT_SCOPED_TASK(itt::domains::MKLDNNPlugin, node->profiling.execute);

        node->execute(s);
    };

    if (config.collectPerfCounters)
        execute(node, stream);
    else
        executeAndMeasure(node, stream);

void MKLDNNGraph::Infer(MKLDNNInferRequest* request, int batch) {
    ...
    for (const auto& node : graphNodes) {
          ...
          executeNode(node, stream);
         ...
    }
}

Most of the function and lambda calls will be inlined according to the assembly.
}

```

**Comment by maxnick** [comment]
> > > I just remembered yet another important thing that I missed earlier. In the previous implementation, we allocated `PerfHelper` object on stack with minimal additional overhead on allocation (just rsp decrease), but now, due to the use of the unique ptr, we allocate the `PerfHelper` object on the heap, which means additional malloc call on every iteration, which significantly increases the overhead cost versus stack allocation. The question is whether we can afford such additional overhead for time measurement in a real application, because although malloc is more expensive than stack allocation, it can still be negligible compared to the entire node execution procedure.
> > 
> > 
> > @maxnick good catch, thank you! fps drop between case with pc and case without pc is near 5%
> > @dmitry-gorokhov is this overhead ok?
> 
> @maxnick @a-sidorova One of the PRs I am working on is also updating Infer() method.
> Can we do the following instead of unique_ptr?:
> 
> ```c++
> inline void MKLDNNGraph::executeNode(const MKLDNNNodePtr& node, const mkldnn::stream& stream) const {
>     auto execute = [](const MKLDNNNodePtr& node, const mkldnn::stream& s) {
>         OV_ITT_SCOPED_TASK(itt::domains::MKLDNNPlugin, node->profiling.execute);
>         node->execute(s);
>     };
> 
>     auto executeAndMeasure = [](const MKLDNNNodePtr& node, const mkldnn::stream& s) {
>         PERF(node);
>         OV_ITT_SCOPED_TASK(itt::domains::MKLDNNPlugin, node->profiling.execute);
> 
>         node->execute(s);
>     };
> 
>     if (config.collectPerfCounters)
>         execute(node, stream);
>     else
>         executeAndMeasure(node, stream);
> 
> void MKLDNNGraph::Infer(MKLDNNInferRequest* request, int batch) {
>     ...
>     for (const auto& node : graphNodes) {
>           ...
>           executeNode(node, stream);
>          ...
>     }
> }
> 
> Most of the function and lambda calls will be inlined according to the assembly.
> }
> ```

Of course. But why do we need this lambdas? We can just inline the code.

**Comment by EgorDuplensky** [comment]
> Of course. But why do we need this lambdas? We can just inline the code.

@maxnick Lambdas will be inlined anyway, so this for readability as usual.

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.h (hunk 8 above)
Comment (a-sidorova) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.h]: We want to remove useless code in inference time and add it to load time to avoid overhead during inference.
If we would pass nodes to functions as parameters without class fields, we should extract constant nodes before `Infer` method but inside `InferImpl`. But it's inference anyway. So we won't have any improvements
So I just added comments for class fields

The nodes are shared_ptr, so we have some sync between graphNodes and constantGraphNodes, mutableGraphNodes (unfortunately, this is not about cases of adding or removing nodes)

Comment (maxnick) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.h]: As far as I understand we call `ExtractConstantNodes()` after all possible transformations at the end of the `InitGraph()` procedure, so the graphNodes content should not be changed afterwards. Therefore such splitting is safe by the internal contract and graph initialization algorithm itself. But, to insure better reliability we can extract the  Replication and Transformation stages into another class. So that the new class become a kind of factory for constant graph object that may not be changed and can perform inference, input/output processing, data damping etc.

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.h (hunk 7 above)
Comment (maxnick) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.h]: When you see `Extract`, you are waiting for the return value. M.b. only in my mind. What would you say about "SeparateConstantNodes"?

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 9]
```diff
@@ -810,29 +819,44 @@ void MKLDNNGraph::Infer(MKLDNNInferRequest* request, int batch) {
...
             request->ThrowIfCanceled();
-        }
 
-        PERF(graphNodes[i]);
+        ENABLE_CPU_DEBUG_CAP(nd.dumpInputBlobs(node));
+        ENABLE_CPU_DEBUG_CAP(nd.dumpOutputBlobs(node));
+    }
+#endif
+
+    for (const auto& node : mutableGraphNodes) {
```
Comment (EgorDuplensky) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:832]: This could actually have also improved performance in theory (foreach uses **!=** comparison vs **<** comparison)
Just wondering, have you noticed some change in performance after this change?

Comment (a-sidorova) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:832]: It is hard to say, at least I didn't notice changes unfortunately

**Review by dmitry-gorokhov — APPROVED** [review]
LGTM. Can be merged after internal validation

**Comment by dmitry-gorokhov** [comment]
@maxnick @EgorDuplensky any additional comments?
@a-sidorova please rebase on latest master

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 10]
```diff
@@ -810,29 +819,44 @@ void MKLDNNGraph::Infer(MKLDNNInferRequest* request, int batch) {
...
-        ENABLE_CPU_DEBUG_CAP(nd.dumpOutputBlobs(graphNodes[i]));
+        ENABLE_CPU_DEBUG_CAP(nd.dumpOutputBlobs(node));
     }
 
     if (infer_count != -1) infer_count++;
 }
 
+inline void MKLDNNGraph::ExecuteNode(const MKLDNNNodePtr& node, const mkldnn::stream& stream) const {
+    if (config.collectPerfCounters) {
+        PERF(node);
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:851]: @a-sidorova  @maxnick @EgorDuplensky still think we need to measure performance of the whole cycle content (including ThrowIfCanceled and etc). This overhead we pay for each node and it is CPU plug-in specific so logically should be included into pc. Also sum of perf counters has implicit requirement to be as much as close to resulting latency measured outside of infer call.

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.h [hunk 11]
```diff
@@ -217,14 +217,21 @@ class MKLDNNGraph {
...
     void CreatePrimitives();
+    void ExtractConstantNodes();
     void ExecuteConstantNodesOnly();
 
     friend class MKLDNNInferRequest;
     friend class MKLDNNGraphlessInferRequest;
     friend InferenceEngine::CNNNetwork dump_graph_as_ie_ngraph_net(const MKLDNNGraph &graph);
 
 private:
+    // these copies of graphNodes are to avoid regular checking for
```
Comment (maxnick) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph.h:228]: Pointers, not copies.

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 10 above)
Comment (maxnick) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:851]: Technically this overhead is a constant addition to the node execution time, which is, with no doubts, important to measure but does not help to understand the node execution time in itself. Moreover, it can even  blur node execution time if the overhead far outweighs  the node time. So practically we need latency measurement, which of course considers this overhead for the whole inference, but also we need precise execution time for each node in itself.

Comment (EgorDuplensky) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:851]: Agree with @maxnick.
Also having node execution time as clean as possible even helps to measure the overhead, as Dmitry mentioned:
_Total Overhead = Latency - SumOfNodeExecutions._

Comment (dmitry-gorokhov) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:851]: As it was stated many times: both metrics are important. The question is what we expose as default number. As soon as perf counters is part of public API and frequently used by customers to get brief performance information I would prefer to include overheads into calculation (actually this is honest numbers which OV takes for each layer execution). 
Task to distinguish between overheads and useful computations is more developer-specific. That's why is to put such counters under DEBUG_CAPS.

Comment (EgorDuplensky) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:851]: From my perspective measurement of per layer execution time is already developer-specific and Latency is for the customers.
But maybe I am wrong.
Anyway, Dmitry is the one to decide after having all the arguments provided :)

Comment (dmitry-gorokhov) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:851]: Than the decision is to measure whole loop content by default. Overhead is to be measured under DEBUG_CAPS (separate PR). Further discussions lets move offline.

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.h (hunk 11 above)
Comment (a-sidorova) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.h]: Yes, thanks. But I left information about graphNodes because these ptrs are equal and connected

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp (hunk 10 above)
Comment (a-sidorova) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp]: Well, I added measure whole loop content using `unique_ptr`. And I didn't notice perf regression due to the creation of `unique_ptr`

**Review by dmitry-gorokhov — APPROVED** [review]

**Comment by dmitry-gorokhov** [comment]
@maxnick  @EgorDuplensky any additional comments?

**Review by dmitry-gorokhov — APPROVED** [review]
